src/physics/grid/ValueGridStore.cc

```diff
@@ -0,0 +1,102 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2021 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file ValueGridStore.cc
+//---------------------------------------------------------------------------//
+#include "ValueGridStore.hh"
+
+#include "base/SpanRemapper.hh"
+#include "base/VectorUtils.hh"
+#include "comm/Device.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with storage space requirements.
+ *
+ * This allocates the required amount of data on the host so that the vector
+ * can keep track of its actual size.
+ */
+ValueGridStore::ValueGridStore(size_type num_grids, size_type num_values)
+    : capacity_(num_grids)
+{
+    CELER_EXPECT(num_grids > 0);
+    CELER_EXPECT(num_values > 0);
+    host_values_.reserve(num_values);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Add a grid of host pointer data.
+ */
+void ValueGridStore::push_back(const XsGridPointers& inp)
+{
+    CELER_EXPECT(inp);
+    CELER_EXPECT(this->size() != this->capacity());
+    CELER_EXPECT(host_values_.size() + inp.value.size()
+                 <= host_values_.capacity());
+
+    host_xsgrids_.push_back(inp);
+    host_xsgrids_.back().value = celeritas::extend(inp.value, &host_values_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Add a grid of host pointer data.
+ */
+void ValueGridStore::push_back(const GenericGridPointers&)
+{
+    CELER_NOT_IMPLEMENTED("generic grids");
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Copy all data to device.
+ */
+void ValueGridStore::copy_to_device()
```

Comment:
Shouldn't this series of function (optionally?) take a (cuda) stream to allow for asynchronous uploads?

Reply:
None of the copying  or kernel launches are yet asynchronous. When the time comes to add streams and other asynchronous aspects we may not want to require a stream argument to be passed everywhere. Besides, if I recall correctly, asynchronous copies usually require special CUDA-allocated pinned memory, which the `std::vector` backend does not do.

Regardless of asynchronicity, in this particular case I don't think it will matter since this should only be called once (per class use anyway) during setup, and not at all during runtime.

Reply:
Fair enough (too soon and need thought/design).  On the last point, even for the one time initialization it might save some (significant?) startup-time to overlap reading and preparing the (many) tables and loading them on the GPU (it all depends on the number and size and cost of any preprocessing of those tables).

Reply:
Yep. Though comparing against the VecGeom load time for CMS... 😉